README.md
[![Build Status](https://travis-ci.org/justanotherdot/attend.svg?branch=master)](https://travis-ci.org/justanotherdot/attend)

# [attend][]

Thanks for starting a project with Haskeleton! If you haven't heard of it
before, I suggest reading the introductory blog post. You can find it here:
<http://taylor.fausak.me/2014/03/04/haskeleton-a-haskell-project-skeleton/>.

Before you get started, there are a few things that this template couldn't
provide for you. You should:

-   Add a synopsis to `package.yaml`. It should be a short (one sentence)
    explanation of your project.

-   Add a description to `package.yaml`. This can be whatever you want it to
    be.

-   Add a category to `package.yaml`. A list of categories is available on
    Hackage at <http://hackage.haskell.org/packages>.

-   Rename `library/Example.hs` to whatever you want your top-level module to
    be called. Typically this is the same as your package name but in
    `CamelCase` instead of `kebab-case`.

    -   Don't forget to rename the reference to it in
        `executable/Main.hs`!

-   If you are on an older version of Stack (<1.0.4), delete `package.yaml` and
    remove `/*.cabal` from your `.gitignore`.

Once you've done that, start working on your project with the Stack commands
you know and love.

``` sh
# Build the project.
stack build

# Run the test suite.
stack test

# Run the benchmarks.
stack bench

# Generate documentation.
stack haddock
```

Thanks again, and happy hacking!

[attend]: https://github.com/justanotherdot/attend


CHANGELOG.md
# Change log

attend uses [Semantic Versioning][].
The change log is available through the [releases on GitHub][].

[Semantic Versioning]: http://semver.org/spec/v2.0.0.html
[releases on GitHub]: https://github.com/justanotherdot/attend/releases


test/Spec.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes       #-}
module Main (main) where

{-import           Lib                 (app)-}
{-import           Test.Hspec-}
{-import           Test.Hspec.Wai-}
{-import           Test.Hspec.Wai.JSON-}

main :: IO ()
main = putStrLn "No tests implemented yet"

{-main :: IO ()-}
{-main = hspec spec-}

{-spec :: Spec-}
{-spec = describe "Routes" $ with (return app) $ do-}
  {-context "/" $ do-}
    {-it "responds with 200" $ do-}
      {-get "" `shouldRespondWith` 200-}
      {-get "/" `shouldRespondWith` 200-}


library/Example.hs
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}

module Example (startApp) where

import           Network.HTTP.Types.Status      (status404)
import           Network.Wai                    (responseFile)
import           Network.Wai.Application.Static (defaultWebAppSettings)
import           Network.Wai.Handler.Warp       (defaultSettings, setPort,
                                                 setTimeout)
import           Network.Wai.Handler.WarpTLS    (runTLS, tlsSettings)
import           Servant                        (Application, Proxy (..), Raw,
                                                 Server, serve,
                                                 serveDirectoryWith)
import           WaiAppStatic.Types             (StaticSettings (..),
                                                 unsafeToPiece)

type API = Raw

certLoc :: FilePath
certLoc = "certificate.pem"

keyLoc :: FilePath
keyLoc = "key.pem"

-- When running in production set the port explicitly to 443
startApp :: IO ()
startApp = runTLS tlsConf serverConf app
  where
    tlsConf    = tlsSettings certLoc keyLoc
    serverConf = setTimeout 15 . setPort 3000 $ defaultSettings

app :: Application
app = serve api server

api :: Proxy API
api = Proxy

server :: Server API
server = serveDirectoryWith settings
  where
    dfs = defaultWebAppSettings rootPath
    ixs  = map unsafeToPiece ["index.html", "index.htm"]
    settings = dfs{ ssIndices    = ixs
                  , ss404Handler = Just custom404
                  , ssUseHash    = False
                  }

rootPath :: FilePath
rootPath = "/var/www"

html404 :: FilePath
html404 = "404.html"

custom404 :: Application
custom404 _ sendResponse = sendResponse response
  where
    headers = [("Content-Type", "text/html; charset=UTF-8")]
    response = responseFile status404 headers html404 Nothing


src/Lib.hs
{-# LANGUAGE DataKinds         #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators     #-}

module Lib
    ( startApp
    , app
    ) where

import           Network.HTTP.Types.Status      (status404)
import           Network.Wai                    (responseFile)
import           Network.Wai.Application.Static (defaultWebAppSettings)
import           Network.Wai.Handler.Warp       (defaultSettings, setPort,
                                                 setTimeout)
import           Network.Wai.Handler.WarpTLS    (runTLS, tlsSettings)
import           Servant                        (Application, Proxy (..), Raw,
                                                 Server, serve,
                                                 serveDirectoryWith)
import           WaiAppStatic.Types             (StaticSettings (..),
                                                 unsafeToPiece)

type API = Raw

certLoc :: FilePath
certLoc = "certificate.pem"

keyLoc :: FilePath
keyLoc = "key.pem"

-- When running in production set the port explicitly to 443
startApp :: IO ()
startApp = runTLS tlsConf serverConf app
  where
    tlsConf    = tlsSettings certLoc keyLoc
    serverConf = setTimeout 15 . setPort 3000 $ defaultSettings

app :: Application
app = serve api server

api :: Proxy API
api = Proxy

server :: Server API
server = serveDirectoryWith settings
  where
    dfs = defaultWebAppSettings rootPath
    ixs  = map unsafeToPiece ["index.html", "index.htm"]
    settings = dfs{ ssIndices    = ixs
                  , ss404Handler = Just custom404
                  , ssUseHash    = False
                  }

rootPath :: FilePath
rootPath = "/var/www"

html404 :: FilePath
html404 = "404.html"

custom404 :: Application
custom404 _ sendResponse = sendResponse response
  where
    headers = [("Content-Type", "text/html; charset=UTF-8")]
    response = responseFile status404 headers html404 Nothing
